smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/transactions/KafkaTransactionsImpl.java

```diff
@@ -0,0 +1,192 @@
+            return new Transaction<R>().execute(work);
+        }
+        throw KafkaExceptions.ex.transactionInProgress(name);
+    }
+
+    @SuppressWarnings("rawtypes")
+    @Override
+    @CheckReturnValue
+    public synchronized <R> Uni<R> withTransaction(Message<?> message, Function<TransactionalEmitter<T>, Uni<R>> work) {
+        String channel;
+        Map<TopicPartition, OffsetAndMetadata> offsets;
+
+        Optional<IncomingKafkaRecordBatchMetadata> batchMetadata = message.getMetadata(IncomingKafkaRecordBatchMetadata.class);
+        Optional<IncomingKafkaRecordMetadata> recordMetadata = message.getMetadata(IncomingKafkaRecordMetadata.class);
+        if (batchMetadata.isPresent()) {
+            IncomingKafkaRecordBatchMetadata<?, ?> metadata = batchMetadata.get();
+            channel = metadata.getChannel();
+            offsets = metadata.getOffsets().entrySet().stream()
+                    .collect(Collectors.toMap(Map.Entry::getKey, e -> new OffsetAndMetadata(e.getValue().offset() + 1)));
+        } else if (recordMetadata.isPresent()) {
+            IncomingKafkaRecordMetadata<?, ?> metadata = recordMetadata.get();
+            channel = metadata.getChannel();
+            offsets = new HashMap<>();
+            offsets.put(new TopicPartition(metadata.getTopic(), metadata.getPartition()),
+                    new OffsetAndMetadata(metadata.getOffset() + 1));
+        } else {
+            throw KafkaExceptions.ex.noKafkaMetadataFound(message);
+        }
+
+        KafkaConsumer<?, ?> consumer = clientService.getConsumer(channel);
+        if (consumer == null) {
+            throw KafkaExceptions.ex.unableToFindConsumerForChannel(channel);
+        }
+        if (currentTransaction == null) {
+            return new Transaction<R>(
+                    /* before commit */
+                    consumer.consumerGroupMetadata()
+                            .chain(groupMetadata -> producer.sendOffsetsToTransaction(offsets, groupMetadata)),
+                    r -> Uni.createFrom().item(r),
+                    VOID_UNI,
+                    /* after abort */
+                    t -> consumer.resetToLastCommittedPositions()
+                            .chain(() -> Uni.createFrom().failure(t))).execute(work);
+        }
+        throw KafkaExceptions.ex.transactionInProgress(name);
+    }
+
+    private static final Uni<Void> VOID_UNI = Uni.createFrom().voidItem();
+
+    private static <R> Uni<R> defaultAfterCommit(R result) {
+        return Uni.createFrom().item(result);
+    }
+
+    private static <R> Uni<R> defaultAfterAbort(Throwable throwable) {
+        return Uni.createFrom().failure(throwable);
+    }
+
+    private class Transaction<R> implements TransactionalEmitter<T> {
+
+        private final Uni<Void> beforeCommit;
+        private final Function<R, Uni<R>> afterCommit;
+
+        private final Uni<Void> beforeAbort;
+        private final Function<Throwable, Uni<R>> afterAbort;
+
+        boolean abort;
```

Comment:
Should this be `volatile` in case `withTransaction` wasn't called on a Vert.x context?

(Genuine question. I'm not sure what executes on the sending thread and what on the original thread that called `withTransaction`.)

Reply:
I remember asking myself the same question for `abort`. But I don't know why I didn't make it volatile. 

So for the execution threads: 
If `withTransaction` is NOT called on Vert.x context, all pipeline, including the execution of the passed function will run on the producer sender thread. 
If it is called on Vert.x context, call the passed function on that context, but producer operations will run on the producer sender thread.

So yes if called on Vert.x context `abort` can be accessed by both the sender thread and the Vert.x thread.